### ISSUE TITLE
BTreeMap online compaction

### ISSUE THREAD

**Issue by jankotek** [jankotek/mapdb#97]
MapDB uses B-Linked-Tree described by Lehman and Yao. This is nearly lock-free concurrent structure. Trade-off is that empty tree nodes are not released after deletes. Instead empty nodes are left behind after delete. This may become problem for long-term use. 

Current workaround for this fragmentation is to recreate BTreeMap (create new, insert all data, replace original). However there is compaction algorithm which requires minimal locking and can work while tree is online. It is described here:
http://www.doc.ic.ac.uk/~td202/

MapDB should implement this compaction and make it optionally running on back-round thread.

**Comment by freakolowsky** [comment]
Would it be possible to add another version of DB.compact() (i.e. cloneCompact()) method that would do exactly what the workaround suggests(create new, copy all collections, overwrite the old state file) without having to close the database(just block all operations while cloning is being done)?

Currently i have to do shutdown the process and do that manually.

**Comment by jankotek** [comment]
@freakolowsky that is doable. I could lock all segments for read and recreate database. Original db would still be readable. 

Also MapDB 1.0 already has snapshots, which could be used for this.

**Comment by freakolowsky** [comment]
but aren't snapshots read only? ...

**Comment by jankotek** [comment]
Yes they are. But one could take snapshot and insert its content to other database.

**Comment by freakolowsky** [comment]
That i understand.

I could:
- create a snapshot
- clear my database
- reinsert everything from snapshot

But i have a few other questions:
- i'm using async writer, how do pause it and make sure it has no more pending operations, i can guarantee that no more new operations will be performed on the target database
- iz a snaphsot a duplicate of a real database or is it a clone copy? does a snapshot retain same fragmentation as original. if it does not have the fragmentation as does the original, could i somehow replace original with the snapshot online?
- how do i clear my database(or collection) so that it's really clear, as if it was just created

**Comment by jankotek** [comment]
> i'm using async writer, how do pause it and make sure it has no more pending operations, i can guarantee that no more new operations will be performed on the target database

db.getEngine().clearCache() flushes the write queue

> iz a snaphsot a duplicate of a real database or is it a clone copy? 

It is original with copy on write. Old values are kept on heap

> how do i clear my database(or collection) so that it's really clear, as if it was just created

Delete files and create new  database.

**Comment by jankotek** [comment]
Duplicate of https://github.com/jankotek/mapdb/issues/545
